src/main/java/com/google/sps/servlets/UserDataServlet.java

```diff
@@ -15,18 +16,142 @@
+import com.google.gson.Gson;
 
-/** Servlet to add user profile data to the Datastore */
-@WebServlet("/data")
+/** Servlet that provides information about a specific user, and allows setting a user's info */
+@WebServlet("/user-data")
 public class UserDataServlet extends HttpServlet {
+  private static final String DEFAULT_STRING = "";
+  private static final String USER_ENTITY = "User";
+  private static final String USER_BIO_PROPERTY = "bio";
+  private static final String USER_EMAIL_PROPERTY = "email";
+  private static final String USER_FOUND_PROPERTY = "user-found";
+  private static final String USER_FRIENDS_LIST_PROPERTY = "friends-list";
+  private static final String USER_ID_PROPERTY = "id";
+  private static final String USER_NAME_PROPERTY = "name";
+
+  private DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  private final Gson gson = new Gson();
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html");
-    response.getWriter().println("Sample response.");
+    // Get the userId, and query Datastore for the corresponding entity
+    String userId = getStringParameter(request, USER_ID_PROPERTY, DEFAULT_STRING);
+    PreparedQuery results = datastore.prepare(new Query(userId));
+
+    Map<String, Object> userData = new HashMap<>();
+    if (results.countEntities() == 0) {
+      // If a user entity was not found
+      userData.put(USER_FOUND_PROPERTY, false);
+    }
+    else {
+      // If a user entity was found (a single entity)
+      Entity userEntity = results.asSingleEntity();
+
+      // Get the user's information
+      String name = (String) userEntity.getProperty(USER_NAME_PROPERTY);
+      String id = (String) userEntity.getProperty(USER_ID_PROPERTY);
+      String email = (String) userEntity.getProperty(USER_EMAIL_PROPERTY);
+      String bio = (String) userEntity.getProperty(USER_BIO_PROPERTY);
+      ArrayList<String> friendsIds = (ArrayList<String>) userEntity.getProperty(USER_FRIENDS_LIST_PROPERTY);
+
+      userData.put(USER_FOUND_PROPERTY, true);
+      userData.put(USER_NAME_PROPERTY, name);
+      userData.put(USER_ID_PROPERTY, id);
+      userData.put(USER_EMAIL_PROPERTY, email);
+      userData.put(USER_BIO_PROPERTY, bio);
+      userData.put(USER_FRIENDS_LIST_PROPERTY, friendsIds);
+    }
+
+    // Send the user's json data as the response
+    response.setContentType("application/json");
+    response.getWriter().println(gson.toJson(userData));
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the user's info to either update or create
+    String userId = getStringParameter(request, USER_ID_PROPERTY, DEFAULT_STRING);
+    String userName = getStringParameter(request, USER_NAME_PROPERTY, DEFAULT_STRING);
+    String userEmail = getStringParameter(request, USER_EMAIL_PROPERTY, DEFAULT_STRING);
```

Comment:
is the email in the form or is this from the facebook api?

Reply:
From the facebook API. If we wanted to add it to the form later, we could, but right now the user only provides their email when they log-in.